### PR TITLE
Bug fix and refactoring

### PR DESCRIPTION
**Pull request by nitinprabhu**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RPE-574

### Change description ###

- Fix document parsing issue.

- Extracted document processing to a different class.

- Updated integration test cases.

- Added Enable Scheduling annotation.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -173,8 +173,6 @@ dependencies {
 
   compile group: 'uk.gov.hmcts.reform', name: 'service-auth-provider-client', version: '1.0.0'
 
-  compile group: 'org.springframework.hateoas', name: 'spring-hateoas', version: versions.springHateoas
```
Comment (doncem) [inline build.gradle]: 👏 

didn't you add milestone repository for this one too? best to remove if so 🙂

doesn't this mean you can remove `versions.springHateoas` too?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessor.java [hunk 2]
```diff
@@ -118,24 +112,10 @@ private void processZipFile(CloudBlobContainer container, String zipFilename)
             List<ScannableItem> scannedItems = processMetaFile(metadataStream);
 
             // TODO check scannedItems.size == pdfFiles.size
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessor.java]: this todo possibly can be refactored to document processor too, am i right?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/DocumentProcessor.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        scannedItems.forEach(item -> {
+            if (response.containsKey(item.getFileName())) {
+                item.setDocumentUrl(response.get(item.getFileName()));
+            } else {
+                //throw some exception here
+                log.error("Document url not found for document {}", item.getFileName());
+            }
+        });
+
+        scannableItemRepository.saveAll(scannedItems);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/DocumentProcessor.java]: i'd rather keep those items as fire and forget as seen throughout various java practices / implementations. any of this typical modification is preferred to have separate private list to which you add elements. could be overthinking and this approach is good enough 👍

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        DocumentManagementService documentManagementService,
+        ScannableItemRepository scannableItemRepository
+    ) {
+        this.documentManagementService = documentManagementService;
+        this.scannableItemRepository = scannableItemRepository;
+    }
+
+    @Transactional
+    public void processPdfFiles(List<Pdf> pdfs, List<ScannableItem> scannedItems) {
+        // TODO check scannedItems.size == pdfFiles.size
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java]: Should we care when such thing happen? Or should we put an assertion on that?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        Map<String, String> response = documentManagementService.uploadDocuments(pdfs);
+
+        log.info("Document service response with file name and doc url {}", response);
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java]: This looks rather like a debug level to me.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        scannedItems.forEach(item -> {
+            if (response.containsKey(item.getFileName())) {
+                item.setDocumentUrl(response.get(item.getFileName()));
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java]: To be uber-squaky-clean, this is where I would add an info entry:
```java
log.info("{} was stored under {}". item.getFileName(), item.getDocumentUrl());
```

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            } else {
+                throw new DocumentNotFoundException("Document metadata not found for file " + item.getFileName());
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java]: ... and here:
```java
log.warn("{} did not get its DM URL", item.getFileName());
```

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    public EnvelopeProcessor(EnvelopeRepository envelopeRepository) {
+        this.envelopeRepository = envelopeRepository;
+    }
+
+    @Transactional
+    public List<ScannableItem> processEnvelope(byte[] metadataStream) throws IOException {
+        if (Objects.isNull(metadataStream)) {
+            throw new MetadataNotFoundException("No metadata file found in the zip file");
+        }
+        //TODO Perform json schema validation for the metadata file
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: I would leave schema validation to `EntityParser`.

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/tasks/processor/DocumentProcessorTest.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        //then
+        assertThatThrownBy(() -> documentProcessor.processPdfFiles(pdfs, ImmutableList.of(scannableItem)))
+            .isInstanceOf(DocumentNotFoundException.class)
+            .hasMessage("Document metadata not found for file test1.pdf");
+
+        verify(documentManagementService).uploadDocuments(pdfs);
+        verify(scannableItemRepository, times(0)).saveAll(anyCollection());
+    }
+
+    private ScannableItem scannableItem() {
```
Comment (KarolJastrzebski) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/tasks/processor/DocumentProcessorTest.java]: Good candidate for private static const :tipping_hand_man:

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/tasks/processor/DocumentProcessorTest.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        Map<String, String> response = ImmutableMap.of("test1.pdf", "http://localhost/documents/5fef5f98-e875-4084-b115-47188bc9066b");
+
+        when(documentManagementService.uploadDocuments(pdfs)).thenReturn(response);
+
+        //when
+        ScannableItem scannableItem = scannableItem();
+        documentProcessor.processPdfFiles(pdfs, ImmutableList.of(scannableItem));
+
+        //then
+        assertThat(scannableItem.getDocumentUrl()).isEqualTo("http://localhost/documents/5fef5f98-e875-4084-b115-47188bc9066b");
```
Comment (KarolJastrzebski) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/tasks/processor/DocumentProcessorTest.java]: This assertion forces implementation to modify input arguments and boils down to:

```java
ScannableItem a = new ScannableItem();
a.setDocumentUrl(null);

documentProcessor.processPdfFiles(pdf, ImmutableList.of(scannableItem));

assertThat(a.getDocumentUrl()).isNotNull();
```

Better would be to test if `scannableItemRepository` was asked for saving the item with the URL provided by DM.

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/tasks/processor/DocumentProcessorTest.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+        verify(documentManagementService).uploadDocuments(pdfs);
+        verify(scannableItemRepository).saveAll(anyCollection());
+    }
+
+    @Test
+    public void should_throw_doc_not_found_exc_when_doc_response_does_not_contain_matching_file_name_and_doc_url()
```
Comment (KarolJastrzebski) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/tasks/processor/DocumentProcessorTest.java]: Yeah... let's talk options for handling missing DM URLs.

File: build.gradle (hunk 1 above)
Comment (nitinprabhu) [reply build.gradle]: Removed now. Milestone is still required.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java (hunk 4 above)
Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java]: Yes we should but not part of this ticket.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java (hunk 5 above)
Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java]: It won't be logged into app insights if we change it to debug.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java (hunk 6 above)
Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java]: Not required we are already logging the response.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/DocumentProcessor.java]: We would need to throw exception and should be picked up in the next run. Can't be a warning as we fail to process  the evidence.

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/tasks/processor/DocumentProcessorTest.java (hunk 10 above)
Comment (nitinprabhu) [reply src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/tasks/processor/DocumentProcessorTest.java]: Good point changes done 👍

**Pull request merged**